### PR TITLE
Feature/update support in nsg

### PR DESCRIPTION
**Pull request by haider-ali-confiz**

### CODE REVIEW

File: lib/fog/azurerm/models/network/network_security_group.rb [hunk 1]
```diff
@@ -48,6 +48,32 @@ def destroy
           service.delete_network_security_group(resource_group, name)
         end
 
+        def update(hash = {})
+          raise('Provided hash can not be empty.') if hash.empty? || hash.nil?
+
+          if !hash[:name].nil? || !hash[:location].nil? || !hash[:resource_group].nil? || !hash[:id].nil?
```
Comment (abbas-sheikh-confiz) [inline lib/fog/azurerm/models/network/network_security_group.rb]: Check for presence of key in hash via hash.key?('name'). This is more explicit and easier to read.

File: lib/fog/azurerm/models/network/network_security_group.rb [hunk 2]
```diff
@@ -48,6 +48,32 @@ def destroy
...
+            raise('You can not modify id:, name:, location: or resource_group:')
```
Comment (abbas-sheikh-confiz) [inline lib/fog/azurerm/models/network/network_security_group.rb]: Update the messaging "You can not update id, name, location or resource_group of network security group"

File: lib/fog/azurerm/models/network/network_security_group.rb [hunk 3]
```diff
@@ -48,6 +48,32 @@ def destroy
...
+          end
+          unless hash[:security_rules].nil?
+            validate_security_rules(hash[:security_rules])
+            merge_attributes(hash)
+            nsg = service.create_network_security_group(resource_group, name, location, security_rules)
```
Comment (abbas-sheikh-confiz) [inline lib/fog/azurerm/models/network/network_security_group.rb]: rename create_network_security_group to "create_or_update_network_security_group".

Can we implement this as update_security_rule?

File: lib/fog/azurerm/models/network/network_security_group.rb [hunk 4]
```diff
@@ -48,6 +48,32 @@ def destroy
...
+            return merge_attributes(Fog::Network::AzureRM::NetworkSecurityGroup.parse(nsg))
+          end
+          raise 'Invalid hash key.'
+        end
+
+        def add_security_rules(resource_group, security_group_name, security_rules)
```
Comment (abbas-sheikh-confiz) [inline lib/fog/azurerm/models/network/network_security_group.rb]: We do not need to pass in resource_group and network_security_group_name, as these are instance variables of model class and directly accessible.

File: lib/fog/azurerm/models/network/network_security_group.rb [hunk 5]
```diff
@@ -48,6 +48,32 @@ def destroy
...
+          validate_security_rules(security_rules)
+          nsg = service.add_security_rules(resource_group, security_group_name, security_rules)
```
Comment (abbas-sheikh-confiz) [inline lib/fog/azurerm/models/network/network_security_group.rb]: Please confirm if add_security_rules method updates the original object or creates a new one

File: lib/fog/azurerm/models/network/network_security_group.rb [hunk 6]
```diff
@@ -48,6 +48,32 @@ def destroy
...
+          merge_attributes(Fog::Network::AzureRM::NetworkSecurityGroup.parse(nsg))
+        end
+
+        def remove_security_rule(resource_group, name, security_rule_name)
```
Comment (abbas-sheikh-confiz) [inline lib/fog/azurerm/models/network/network_security_group.rb]: We do not need to pass in resource_group and network_security_group_name, as these are instance variables of model class and directly accessible.

File: lib/fog/azurerm/requests/network/add_security_rules.rb [hunk 7]
```diff
@@ -0,0 +1,152 @@
+module Fog
+  module Network
+    class AzureRM
+      # Real class for Network Request
+      class Real
+        def add_security_rules(resource_group, name, security_rules)
```
Comment (abbas-sheikh-confiz) [inline lib/fog/azurerm/requests/network/add_security_rules.rb]: rename 'name' to 'security_group_name'

File: lib/fog/azurerm/requests/network/add_security_rules.rb [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+  module Network
+    class AzureRM
+      # Real class for Network Request
+      class Real
+        def add_security_rules(resource_group, name, security_rules)
+          Fog::Logger.debug "Updating security rules in Network Security Group #{name}."
+          promise = @network_client.network_security_groups.get(resource_group, name)
+          result = promise.value!
+          nsg = result.body
+          security_rules_list = define_security_rules(security_rules)
```
Comment (abbas-sheikh-confiz) [inline lib/fog/azurerm/requests/network/add_security_rules.rb]: rename 
- security_rules_list to security_rules_objects
- rename sr to security_rule
- “define_security_rules” to create_security_rule_objects

File: lib/fog/azurerm/requests/network/get_network_security_group.rb [hunk 9]
```diff
@@ -0,0 +1,161 @@
+module Fog
+  module Network
+    class AzureRM
+      # Real class for Network Security Group Request
+      class Real
+        def get_network_security_group(resource_group, name)
```
Comment (abbas-sheikh-confiz) [inline lib/fog/azurerm/requests/network/get_network_security_group.rb]: rename
- name to network_security_group

File: lib/fog/azurerm/requests/network/remove_security_rule.rb [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+  module Network
+    class AzureRM
+      # Real class for Network Request
+      class Real
+        def remove_security_rule(resource_group, security_group_name, security_rule_name)
+          Fog::Logger.debug "Deleting security rule #{security_rule_name} from Network Security Group #{security_group_name}."
+          promise = @network_client.network_security_groups.get(resource_group, security_group_name)
+          result = promise.value!
+          nsg = result.body
+          updated_security_rule_list = remove_security_rule_from_list(nsg.properties.security_rules, security_rule_name)
```
Comment (abbas-sheikh-confiz) [inline lib/fog/azurerm/requests/network/remove_security_rule.rb]: Rename
- updated_security_rule_list to updated_security_rules

File: lib/fog/azurerm/requests/network/remove_security_rule.rb [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+        def remove_security_rule_from_list(security_rule_list, rule_name)
+          unless security_rule_list.nil?
+            return security_rule_list.select { |sr| sr.name != rule_name }
+          end
+        end
+      end
+
+      # Mock class for Network Request
+      class Mock
+        def remove_security_rule(resource_group, name, _security_rules)
```
Comment (abbas-sheikh-confiz) [inline lib/fog/azurerm/requests/network/remove_security_rule.rb]: - Rename name to security_group_name

File: lib/fog/azurerm/models/network/network_security_group.rb (hunk 1 above)
Comment (haider-ali-confiz) [inline lib/fog/azurerm/models/network/network_security_group.rb]: I've replaced this method name with `update_security_rule`. So now we don't need to put these checks in it.

File: lib/fog/azurerm/models/network/network_security_group.rb (hunk 2 above)
Comment (haider-ali-confiz) [inline lib/fog/azurerm/models/network/network_security_group.rb]: I've replaced this method name with update_security_rule. So now we don't need to put these checks in it.

File: lib/fog/azurerm/models/network/network_security_group.rb (hunk 3 above)
Comment (haider-ali-confiz) [inline lib/fog/azurerm/models/network/network_security_group.rb]: Done.

File: lib/fog/azurerm/models/network/network_security_group.rb (hunk 4 above)
Comment (haider-ali-confiz) [inline lib/fog/azurerm/models/network/network_security_group.rb]: Yes you were right. Removed unnecessary parameters.

File: lib/fog/azurerm/models/network/network_security_group.rb (hunk 5 above)
Comment (haider-ali-confiz) [inline lib/fog/azurerm/models/network/network_security_group.rb]: `add_security_rules` updates the current object.

File: lib/fog/azurerm/models/network/network_security_group.rb (hunk 6 above)
Comment (haider-ali-confiz) [inline lib/fog/azurerm/models/network/network_security_group.rb]: Done.

File: lib/fog/azurerm/requests/network/add_security_rules.rb (hunk 7 above)
Comment (haider-ali-confiz) [inline lib/fog/azurerm/requests/network/add_security_rules.rb]: Done.

File: lib/fog/azurerm/requests/network/add_security_rules.rb (hunk 8 above)
Comment (haider-ali-confiz) [inline lib/fog/azurerm/requests/network/add_security_rules.rb]: Done

File: lib/fog/azurerm/requests/network/get_network_security_group.rb (hunk 9 above)
Comment (haider-ali-confiz) [inline lib/fog/azurerm/requests/network/get_network_security_group.rb]: I've renamed it to security_group_name.

File: lib/fog/azurerm/requests/network/remove_security_rule.rb (hunk 11 above)
Comment (haider-ali-confiz) [inline lib/fog/azurerm/requests/network/remove_security_rule.rb]: Done.

**Pull request merged**